js/noms/src/path.js

```diff
@@ -382,6 +414,20 @@ export class HashIndexPath {
   }
 }
 
+/**
+ * TypeAnnotation is a PathPart annotation to resolve to the type of the value
+ * it's resolved in.
+ */
+class TypeAnnotation {
+  resolve(v: Value): Promise<Value | null> {
+    return Promise.resolve(getTypeOfValue(v));
```

Comment:
It was like this because TypeAnnotation needs to implement the Part interface. But it seems happy with the tighter type, so cool.